use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,224 @@
+                doMeetingsAsSoonAsPossible(constraintFactory),
+                oneBreakBetweenConsecutiveMeetings(constraintFactory),
+                overlappingMeetings(constraintFactory),
+                assignLargerRoomsFirst(constraintFactory),
+                roomStability(constraintFactory)
+        };
+    }
+
+    // ************************************************************************
+    // Hard constraints
+    // ************************************************************************
+
+    public Constraint roomConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEachUniquePair(MeetingAssignment.class,
+                equal(MeetingAssignment::getRoom),
+                overlapping(assignment -> assignment.getStartingTimeGrain().getGrainIndex(),
+                        assignment -> assignment.getStartingTimeGrain().getGrainIndex() +
+                                assignment.getMeeting().getDurationInGrains()))
+                .penalizeConfigurable((leftAssignment, rightAssignment) -> rightAssignment.calculateOverlap(leftAssignment))
+                .asConstraint("Room conflict");
+    }
+
+    public Constraint avoidOvertime(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEachIncludingUnassigned(MeetingAssignment.class)
+                .filter(meetingAssignment -> meetingAssignment.getStartingTimeGrain() != null)
+                .ifNotExists(TimeGrain.class,
+                        equal(MeetingAssignment::getLastTimeGrainIndex, TimeGrain::getGrainIndex))
+                .penalizeConfigurable(MeetingAssignment::getLastTimeGrainIndex)
+                .asConstraint("Don't go in overtime");
+    }
+
+    public Constraint requiredAttendanceConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEachUniquePair(RequiredAttendance.class,
+                equal(RequiredAttendance::getPerson))
+                .join(MeetingAssignment.class,
+                        equal((leftRequiredAttendance, rightRequiredAttendance) -> leftRequiredAttendance.getMeeting(),
+                                MeetingAssignment::getMeeting))
+                .join(MeetingAssignment.class,
+                        equal((leftRequiredAttendance, rightRequiredAttendance, leftAssignment) -> rightRequiredAttendance
+                                .getMeeting(),
+                                MeetingAssignment::getMeeting),
+                        overlapping((attendee1, attendee2, assignment) -> assignment.getStartingTimeGrain().getGrainIndex(),
+                                (attendee1, attendee2, assignment) -> assignment.getStartingTimeGrain().getGrainIndex() +
+                                        assignment.getMeeting().getDurationInGrains(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex(),
+                                assignment -> assignment.getStartingTimeGrain().getGrainIndex() +
```

Comment:
the grainIndex+durationInGrains is repeated so often that I think it would warrant a getter on assignment.